botocore/endpoints_v2.py

```diff
@@ -0,0 +1,1123 @@
+        for it in the rule's scoped variables. Finally, retrieve the function
+        and call it with the resolved arguments. If an `assign` parameter is
+        present, add the returned value to the rule's scoped variables for
+        the next condition/subordinate rule.
+
+        :type func_signature: dict
+        :param func_signature: A function object such as
+        {
+            "fn": "not",
+            "argv": [
+                {
+                    "fn": "isSet",
+                    "argv": [
+                        {
+                            "fn": "parseURL",
+                            "argv": [
+                                {
+                                    "ref": "Endpoint"
+                                }
+                            ]
+                        }
+                    ]
+                }
+            ]
+        }
+
+        :type scope_vars: dict
+        :param scope_vars: A dictionary combining input parameters and variables
+        that have been scoped throughout the rule's evaluation.
+
+        :returns: The result of the function call(s).
+        """
+        func_args = [
+            self.resolve_value(arg, scope_vars)
+            for arg in func_signature["argv"]
+        ]
+        func_name = self.convert_func_name(func_signature["fn"])
+        func = getattr(self, func_name)
+        result = func(*func_args)
+        if "assign" in func_signature:
+            assign = func_signature["assign"]
+            if assign in scope_vars:
+                raise RuleSetDefinitionError(
+                    msg=f"Assignment {assign} already exists in "
+                    "scoped variables and cannot be overwritten"
+                )
+            scope_vars[assign] = result
+        return result
+
+    def resolve_value(self, value, scope_vars):
+        """Given a value, check if it's a function, reference or template.
+        Return the resolved value.
+
+        :type value: Any
+        :param value: A value that may or may not be a function, reference
+        or template.
+
+        :type scope_vars: dict
+        :param scope_vars: A dictionary combining input parameters and variables
+        that have been scoped throughout the rule's evaluation.
+
+        :returns: A fully resolved value.
+        """
+
+        if self.is_func(value):
+            return self.call_function(value, scope_vars)
+
+        elif self.is_ref(value):
+            return scope_vars.get(value["ref"])
+
+        elif self.is_template(value):
+            return self.resolve_template_string(value, scope_vars)
+
+        return value
+
+    def evaluate_conditions(self, scope_vars):
+        """Determine if all conditions in a rule are not None or False.
+
+        :type scope_vars: dict
+        :param scope_vars: A dictionary combining input parameters and variables
+        that have been scoped throughout the rule's evaluation.
```

Comment:
`scope_vars` as a concept we should try to define once. It's an overarching concept that we can hopefully avoid redocumenting on every method.

Reply:
I only have `type` and `rtype` in all doc strings now